Build on github fails due to old testcontainers version

Github actions for repositories that use ETC fail, with:
```
Caused by: org.testcontainers.containers.ContainerFetchException: Can't get Docker image: RemoteDockerImage(imageName=exasol/docker-db:7.0.4, imagePullPolicy=DefaultPullPolicy())
Caused by: com.github.dockerjava.api.exception.NotFoundException: 
Status 404: {"message":"No such image: testcontainers/ryuk:0.3.0"}
```
See: https://github.com/testcontainers/testcontainers-java/issues/3574

Fix: Update test containers to 1.15.1

--- a/.travis.yml
+++ b/.travis.yml
@@ -8,8 +8,8 @@ services:
   - docker
 
 env:
-  - DOCKER_DB_VERSION=7.0.4
-  - DOCKER_DB_VERSION=6.2.12
+  - DOCKER_DB_VERSION=7.0.6
+  - DOCKER_DB_VERSION=6.2.13
 
 jdk:
   - openjdk11

--- a/doc/changes/changelog.md
+++ b/doc/changes/changelog.md
@@ -1,5 +1,6 @@
 # Changes
 
+* [3.4.1](changes_3.5.0.md)
 * [3.4.1](changes_3.4.1.md)
 * [3.4.0](changes_3.4.0.md)
 * [3.3.1](changes_3.3.1.md)

--- a/doc/changes/changelog.md
+++ b/doc/changes/changelog.md
@@ -1,6 +1,6 @@
 # Changes
 
-* [3.4.1](changes_3.5.0.md)
+* [3.5.0](changes_3.5.0.md)
 * [3.4.1](changes_3.4.1.md)
 * [3.4.0](changes_3.4.0.md)
 * [3.3.1](changes_3.3.1.md)

--- a/doc/changes/changes_3.5.0.md
+++ b/doc/changes/changes_3.5.0.md
@@ -1,0 +1,29 @@
+# Exasol Test Containers 3.5.0, released 2021-02-09
+
+Code name: Host IP detector
+
+## Summary
+
+In the 3.5.0 release of the Exasol Test Containers, we added a new function `getHostIp()` that allows getting the IP
+address of the host the Exasol docker container runs on. This is useful if you want to connect to a debugger or profiler.
+
+Note that this feature is at the moment only available on Linux.
+
+## Features
+
+* #68: Support `getHostIp()` on Linux
+
+With release 3.5.0 the Exasol Test Container now provides the function `getHostIp()` that lets you acquire the IP address of the host that runs the Exasol `docker-db`. This is for example useful if you have a debugger or profiler running on the host and want to connect to it from inside the Docker container.
+
+We also updated the default `docker-db` image version to 7.0.6.
+
+## Dependency updates
+ 
+### Test Dependency Updates
+
+* Updated `org.junit.jupiter:junit-jupiter-engine:5.7.0` to `5.7.1`
+* Updated `org.junit.jupiter:junit-jupiter-params:5.7.0` to `5.7.1`
+* Updated `org.mockito:mockito-junit-jupiter:3.6.28` to `3.7.7`
+* Updated `org.testcontainers:jdbc:1.15.0` to `1.15.1`
+* Updated `org.testcontainers:junit-jupiter:1.15.0` to `1.15.1`
+* Updated `org.testcontainers:testcontainers:1.15.0` to `1.15.1`

--- a/doc/changes/changes_3.5.0.md
+++ b/doc/changes/changes_3.5.0.md
@@ -9,13 +9,15 @@ address of the host the Exasol docker container runs on. This is useful if you w
 
 Note that this feature is at the moment only available on Linux.
 
+We also updated the default `docker-db` image version to 7.0.6.
+
 ## Features
 
-* #68: Support `getHostIp()` on Linux
+* #58: Support `getHostIp()` on Linux
 
-With release 3.5.0 the Exasol Test Container now provides the function `getHostIp()` that lets you acquire the IP address of the host that runs the Exasol `docker-db`. This is for example useful if you have a debugger or profiler running on the host and want to connect to it from inside the Docker container.
+## Bug fixes
 
-We also updated the default `docker-db` image version to 7.0.6.
+* 122: Updated base Test Containers versions to 1.15.1 to prevent build failures
 
 ## Dependency updates
  

--- a/doc/design.md
+++ b/doc/design.md
@@ -71,6 +71,10 @@ The `Plugin` allows controlling functions of an EXAoperation plug-in.
 
 The `ExasolDriverManager` is responsible for installing and uninstalling drivers for external data sources.
 
+## `HostIpDetector`
+
+The `HostIpDetector` detects the IP address of the host the container is running on.
+
 # Runtime
 
 This section describes the runtime behavior of the software.
@@ -213,6 +217,17 @@ Covers:
 * `req~shortened-docker-image-references~1`
 
 Needs: impl, utest
+
+###### Host IP Address Detection 
+`dsn~host-ip-address-detection~1`
+
+The `HostIpDetector` scans the Docker network for the Gateway address and uses this as host address.
+
+Covers:
+
+* `req~reading-the-host-ip-address~1`
+
+Needs: impl, itest
 
 ## Database Access
 

--- a/doc/system_requirements.md
+++ b/doc/system_requirements.md
@@ -222,6 +222,21 @@ Covers:
 
 Needs: dsn
 
+#### Reading the Host IP Address
+`req~reading-the-host-ip-address~1`
+
+Integrators can read the IP address of the host the container is running on.
+
+Rationale:
+
+This is useful for reverse socket connections, the kind of which debugging or profiling agents often use.
+
+Covers:
+
+* [feat~docker-based-exasol-instance~1](#docker-based-exasol-instance)
+
+Needs: dsn
+
 ### Database Access
 
 #### JDBC Connection With Administrator Privileges

--- a/pom.xml
+++ b/pom.xml
@@ -2,7 +2,7 @@
   <modelVersion>4.0.0</modelVersion>
   <groupId>com.exasol</groupId>
   <artifactId>exasol-testcontainers</artifactId>
-  <version>3.4.1</version>
+  <version>3.5.0</version>
   <name>Test containers for Exasol on Docker</name>
   <description>This module provides abstraction for generation, startup, shutdown and use of an
         Exasol database running on Docker.
@@ -13,9 +13,9 @@
     <project.reporting.outputEncoding>UTF-8
     </project.reporting.outputEncoding>
     <java.version>11</java.version>
-    <junit.version>5.7.0</junit.version>
+    <junit.version>5.7.1</junit.version>
     <maven.surefire-and-failsafe.version>3.0.0-M3</maven.surefire-and-failsafe.version>
-    <testcontainers.version>1.15.0</testcontainers.version>
+    <testcontainers.version>1.15.1</testcontainers.version>
     <gpg.skip>true</gpg.skip>
     <sonar.coverage.jacoco.xmlReportPaths>target/site/jacoco/jacoco.xml,target/site/jacoco-it/jacoco.xml
     </sonar.coverage.jacoco.xmlReportPaths>
@@ -147,7 +147,7 @@
     <dependency>
       <groupId>org.mockito</groupId>
       <artifactId>mockito-junit-jupiter</artifactId>
-      <version>3.6.28</version>
+      <version>3.7.7</version>
       <scope>test</scope>
     </dependency>
   </dependencies>

--- a/src/main/java/com/exasol/containers/ExasolContainer.java
+++ b/src/main/java/com/exasol/containers/ExasolContainer.java
@@ -704,4 +704,14 @@ public class ExasolContainer<T extends ExasolContainer<T>> extends JdbcDatabaseC
         }
         return this.driverManager;
     }
+
+    /**
+     * Get the IP address of the host running this container.
+     *
+     * @return IP address of the host
+     */
+    public String getHostIp() {
+        final HostIpDetector detector = new HostIpDetector(this);
+        return detector.getHostIp();
+    }
 }

--- a/src/main/java/com/exasol/containers/ExasolContainerConstants.java
+++ b/src/main/java/com/exasol/containers/ExasolContainerConstants.java
@@ -6,7 +6,7 @@ import java.util.Set;
  * Constants for the Exasol Docker containers (like image ID and version).
  */
 public final class ExasolContainerConstants {
-    public static final String EXASOL_DOCKER_IMAGE_VERSION = "7.0.4";
+    public static final String EXASOL_DOCKER_IMAGE_VERSION = "7.0.6";
     public static final String EXASOL_DOCKER_IMAGE_ID = "exasol/docker-db";
     public static final String EXASOL_DOCKER_IMAGE_REFERENCE = EXASOL_DOCKER_IMAGE_ID + ":"
             + EXASOL_DOCKER_IMAGE_VERSION;

--- a/src/main/java/com/exasol/containers/HostIpDetector.java
+++ b/src/main/java/com/exasol/containers/HostIpDetector.java
@@ -1,0 +1,40 @@
+package com.exasol.containers;
+
+import java.util.Map;
+
+import org.testcontainers.containers.Container;
+
+import com.github.dockerjava.api.model.ContainerNetwork;
+import com.github.dockerjava.api.model.NetworkSettings;
+
+/**
+ * Detector for the IP address of the host running the Exasol `docker-db`.
+ */
+public class HostIpDetector {
+    private final Container<? extends Container<?>> container;
+
+    /**
+     * Create a new {@link HostIpDetector} instance.
+     *
+     * @param container container for which's host the IP should be detected
+     */
+    public HostIpDetector(final Container<? extends Container<?>> container) {
+        this.container = container;
+    }
+
+    /**
+     * Get the IP of a host on which a container runs.
+     *
+     * @return host IP
+     */
+    public String getHostIp() {
+        final NetworkSettings networkSettings = this.container.getContainerInfo().getNetworkSettings();
+        final Map<String, ContainerNetwork> networks = networkSettings.getNetworks();
+        if (networks.size() == 0) {
+            throw new IllegalStateException("F-ETC-HID-1: Unable to determine host IP for \""
+                    + this.container.getDockerImageName() + "\" container because the docker network had no entries");
+        } else {
+            return networks.values().iterator().next().getGateway();
+        }
+    }
+}

--- a/src/main/java/com/exasol/containers/HostIpDetector.java
+++ b/src/main/java/com/exasol/containers/HostIpDetector.java
@@ -27,6 +27,7 @@ public class HostIpDetector {
      *
      * @return host IP
      */
+    // [impl->dsn~host-ip-address-detection~1]
     public String getHostIp() {
         final NetworkSettings networkSettings = this.container.getContainerInfo().getNetworkSettings();
         final Map<String, ContainerNetwork> networks = networkSettings.getNetworks();

--- a/src/test/java/com/exasol/containers/HostIpDetectorTest.java
+++ b/src/test/java/com/exasol/containers/HostIpDetectorTest.java
@@ -1,0 +1,29 @@
+package com.exasol.containers;
+
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.Mockito.when;
+
+import java.util.Collections;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.testcontainers.containers.Container;
+
+import com.github.dockerjava.api.command.InspectContainerResponse;
+import com.github.dockerjava.api.model.NetworkSettings;
+
+@ExtendWith(MockitoExtension.class)
+class HostIpDetectorTest {
+    @Test
+    void testGetHostIpThrowsExceptionOnEmptyNetwork(@Mock final Container<? extends Container<?>> container,
+            @Mock final InspectContainerResponse containerInfo, //
+            @Mock final NetworkSettings networkSettings) {
+        when(container.getContainerInfo()).thenReturn(containerInfo);
+        when(containerInfo.getNetworkSettings()).thenReturn(networkSettings);
+        when(networkSettings.getNetworks()).thenReturn(Collections.emptyMap());
+        final HostIpDetector detector = new HostIpDetector(container);
+        assertThrows(IllegalStateException.class, () -> detector.getHostIp());
+    }
+}

--- a/src/test/java/com/exasol/containers/HostIpDetectorTestIT.java
+++ b/src/test/java/com/exasol/containers/HostIpDetectorTestIT.java
@@ -1,0 +1,52 @@
+package com.exasol.containers;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.in;
+import static org.hamcrest.Matchers.not;
+import static org.junit.jupiter.api.Assertions.assertAll;
+
+import java.net.*;
+import java.util.*;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.condition.EnabledOnOs;
+import org.junit.jupiter.api.condition.OS;
+import org.testcontainers.junit.jupiter.Container;
+import org.testcontainers.junit.jupiter.Testcontainers;
+
+@Testcontainers
+class HostIpDetectorTestIT {
+    @Container
+    private static final ExasolContainer<? extends ExasolContainer<?>> EXASOL = new ExasolContainer<>().withReuse(true);
+
+    @Test
+    @EnabledOnOs({ OS.LINUX })
+    void testDetectorOnLinux() throws UnknownHostException, SocketException {
+        final String detectedIp = EXASOL.getHostIp();
+        final List<String> nonLocalHostAddresses = collectHostAddresses();
+        final String localhostIP = InetAddress.getLocalHost().getHostAddress();
+        assertAll(() -> assertThat("Not localhost", detectedIp, not(localhostIP)),
+                () -> assertThat("Detected IP matches one of the non-local host IP", detectedIp,
+                        in(nonLocalHostAddresses)));
+    }
+
+    private List<String> collectHostAddresses() throws SocketException, UnknownHostException {
+        final List<String> hostAddresses = new ArrayList<>();
+        final Enumeration<NetworkInterface> networkInterfaces = NetworkInterface.getNetworkInterfaces();
+        while (networkInterfaces.hasMoreElements()) {
+            final NetworkInterface networkInterface = networkInterfaces.nextElement();
+            collectAllNetworkInterfaceAddresses(networkInterface, hostAddresses);
+        }
+        return hostAddresses;
+    }
+
+    private void collectAllNetworkInterfaceAddresses(final NetworkInterface networkInterface,
+            final List<String> hostAddresses) {
+        final Enumeration<InetAddress> addresses = networkInterface.getInetAddresses();
+        while (addresses.hasMoreElements()) {
+            final InetAddress address = addresses.nextElement();
+            final String foundAddress = address.getHostAddress();
+            hostAddresses.add(foundAddress);
+        }
+    }
+}

--- a/src/test/java/com/exasol/containers/HostIpDetectorTestIT.java
+++ b/src/test/java/com/exasol/containers/HostIpDetectorTestIT.java
@@ -19,6 +19,7 @@ class HostIpDetectorTestIT {
     @Container
     private static final ExasolContainer<? extends ExasolContainer<?>> EXASOL = new ExasolContainer<>().withReuse(true);
 
+    // [itest->dsn~host-ip-address-detection~1]
     @Test
     @EnabledOnOs({ OS.LINUX })
     void testDetectorOnLinux() throws UnknownHostException, SocketException {

